src/Microsoft.Health.Fhir.TemplateManagement/ArtifactProviders/BlobTemplateProvider.cs

```diff
@@ -0,0 +1,90 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using DotLiquid;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Health.Fhir.Liquid.Converter.Utilities;
+using Microsoft.Health.Fhir.TemplateManagement.Models;
+
+namespace Microsoft.Health.Fhir.TemplateManagement.ArtifactProviders
+{
+    public class BlobTemplateProvider : IConvertDataTemplateProvider
+    {
+        private BlobContainerClient _blobContainerClient;
+
+        private readonly IMemoryCache _templateCache;
+
+        private readonly TemplateCollectionConfiguration _templateCollectionConfiguration;
+
+        private readonly string _blobTemplateCacheKey = "cached-blob-templates";
+
+        private readonly int _segmentSize = 100;
+
+        public BlobTemplateProvider(BlobContainerClient blobContainerClient, IMemoryCache templateCache, TemplateCollectionConfiguration templateConfiguration)
+        {
+            _blobContainerClient = blobContainerClient;
+            _templateCache = templateCache;
+            _templateCollectionConfiguration = templateConfiguration;
+        }
+
+        public async Task<List<Dictionary<string, Template>>> GetTemplateCollectionAsync(CancellationToken cancellationToken = default)
+        {
+            // read templates from cache if available
+            if (_templateCache.TryGetValue(_blobTemplateCacheKey, out var templateCache))
```

Comment:
nit: can we use the explicit type (List<Dictionary<string, Template>>) in the out param and skip the cast?